apps/system/js/app_chrome.js

```diff
@@ -340,6 +373,67 @@
     window.dispatchEvent(new CustomEvent('global-search-request'));
   };
 
+  AppChrome.prototype.onClickSiteIcon = function onClickSiteIcon() {
+    this.pinningPref && this.pinningPref.get(function(enabled) {
+      if (!enabled || !this.app.isBrowser()) {
+        this.app.debug('Pinning the web disabled');
+        return;
+      }
+      this.setPinDialogCard();
+    }.bind(this));
+  };
+
+  AppChrome.prototype.setPinDialogCard = function ac_setPinDialogCard(url) {
+    var card = new GaiaPinCard();
+    card.title = this.title.textContent;
+    card.icon = this.siteIcon.style.backgroundImage;
```

Comment:
A slight inconsistency here is that if we want a 32px icon and we get given a 16px icon, the one in the URL bar will have a visual treatment applied to add a circular background. This should happen on the card in the preview and on the homescreen too so that they match. I'm not sure how easy it is for this to follow the same code path?